core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -2826,6 +2826,21 @@ private CorrelationFinder() {
     @Override public Void visitCorrelVariable(RexCorrelVariable var) {
       throw Util.FoundOne.NULL;
     }
+
+    @Override public Void visitSubQuery(RexSubQuery subQuery) {
+      if (!deep) {
+        return null;
+      }
+
+      for (RexNode operand : subQuery.operands) {
+        operand.accept(this);
+      }
+
+      if (!RelOptUtil.getVariablesUsed(subQuery.rel).isEmpty()) {
+        throw Util.FoundOne.NULL;
```

Comment:
From a quick look it seems that `RelOptUtil.getVariablesUsed` on its own does exactly what you want. There is no need to call it from here. 

In fact, there is also `org.apache.calcite.plan.RelOptUtil#notContainsCorrelation` which might be exactly what you are looking for.

Reply:
`RelOptUtil.getVariablesUsed` can collect all the variables used in the Filter operator, including those from the input plan and the filter's Condition. Here, we only need to focus on the variables within the Condition.
`RelOptUtil.notContainsCorrelation` determines whether a specific correlation is included. Here, we need to check whether a correlation exists.
Did I misunderstand something?

Reply:
My bad, I rushed a bit through the code and didn't realize that it was a full-blown RelNode visitor. The changes here do make sense now.